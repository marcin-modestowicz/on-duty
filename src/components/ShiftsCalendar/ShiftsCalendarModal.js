import React, { Component } from "react";
import { observer } from "mobx-react";
import classnames from "classnames";
import {
  AVAILABILITY_STATUSES_NAMES,
  AVAILABILITY_STATUSES_COLORS
} from "../../models/Availability";
import type User from "../../models/User";
import type Shift from "../../models/Shift";
import CheckboxButton from "../CheckboxButton";
import styles from "./ShiftsCalendarModal.scss";

/* global SyntheticInputEvent */

type Props = {
  users: User[],
  day: { date: Date, shift: Shift },
  onClose: () => void
};

@observer
class ShiftsCalendarModal extends Component {
  handleInputChange = (event: SyntheticInputEvent) => {
    const userId = event.target.value;
    const user = this.props.users.find(({ id }) => id === userId);

    if (user) {
      this.props.day.shift.toggleUser(user);
    }
  };

  renderUser(user) {
    const day = this.props.day;
    const userId = user.id;
    const isUserActive = day.shift.onDuty.some(
      activeUser => activeUser.id === userId
    );
    const dayInAvailabilityCalendar = user.availabilityCalendar.days.find(
      ({ date }) => date.toString() === day.date.toString()
    );
    const availabilityStatus = dayInAvailabilityCalendar.availability.status;
    const statusName = AVAILABILITY_STATUSES_NAMES[availabilityStatus];
    const statusColor = AVAILABILITY_STATUSES_COLORS[availabilityStatus];

    return (
      <CheckboxButton
        key={userId}
        id={userId}
        value={userId}
        label={user.name}
        checked={isUserActive}
        onChange={this.handleInputChange}
        color={statusColor}
      />
    );
  }

  render() {
    const { users, onClose } = this.props;

    return (
      <div className={styles.root}>
        <div className={styles.modal}>
          <div className={styles.header}>
            <div role="button" className={styles.button} onClick={onClose}>
              Close
            </div>
          </div>
          <div className={styles.content}>
            {users.map(user => this.renderUser(user))}
          </div>
        </div>
      </div>
    );
  }
}

export default ShiftsCalendarModal;
