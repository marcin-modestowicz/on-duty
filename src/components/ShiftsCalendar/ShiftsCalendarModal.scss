@import '../../styles/colors.scss';

.root {
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: space-around;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,.5);

  .modal {
    min-width: 15em;
    background-color: white;

    .header {
      padding: 1em;

      .button {
        float: right;
        cursor: pointer;
        user-select: none;
      }
    }
    .content {
      padding: 1em;

      .user {
        margin-bottom: .5em;
      }
    }

    .input {
      display: none;
    }

    .label {
      display: block;
      box-sizing: border-box;
      width: 100%;
      height: 2.5em;
      padding: .75em;
      margin: .25em;
      border-width: 0 1px 1px 0;
      border-style: solid;
      border-radius: .25em;
      cursor: pointer;
      user-select: none;
      text-align: center;
      color: white;

      &.free {
        background-color: transparentize($free, 0.8);
        border-color: transparentize($free-darker, 0.9);
      }

      &.keen {
        background-color: transparentize($keen, 0.8);
        border-color: transparentize($keen-darker, 0.9);
      }

      &.busy {
        background-color: transparentize($busy, 0.8);
        border-color: transparentize($busy-darker, 0.9);
      }
    }

    .input:checked + .label {
      border-width: 1px 0 0 1px;

      &.free {
        background-color: $free;
        border-color: $free-darker;
      }

      &.keen {
        background-color: $keen;
        border-color: $keen-darker;
      }

      &.busy {
        background-color: $busy;
        border-color: $busy-darker;
      }
    }
  }
}
