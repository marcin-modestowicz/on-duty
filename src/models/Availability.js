// @flow
import { observable, computed, action } from "mobx";

export const AVAILABILITY_STATUSES = {
  FREE: 0,
  KEEN: 1,
  BUSY: -1
};

export const AVAILABILITY_STATUSES_NAMES = {
  "0": "Free",
  "1": "Keen",
  "-1": "Busy"
};

export type AvailabilityStatus = -1 | 0 | 1;

class Availability {
  @observable status: AvailabilityStatus;

  constructor(availability?: AvailabilityStatus) {
    this.status = availability || AVAILABILITY_STATUSES.FREE;
  }

  @computed
  get statusName(): AvailabilityStatus {
    return AVAILABILITY_STATUSES_NAMES[this.status];
  }

  @action
  toggle() {
    switch (this.status) {
      case AVAILABILITY_STATUSES.BUSY:
        this.status = AVAILABILITY_STATUSES.FREE;
        break;
      case AVAILABILITY_STATUSES.KEEN:
        this.status = AVAILABILITY_STATUSES.BUSY;
        break;
      default:
        this.status = AVAILABILITY_STATUSES.KEEN;
    }
  }

  @action
  setStatus(status: AvailabilityStatus) {
    this.status = status;
  }
}

export default Availability;
