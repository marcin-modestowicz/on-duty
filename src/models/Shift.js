// @flow
import { observable, action, computed } from "mobx";
import type User from "./User";

export const USERS_PER_SHIFT = 2;

export class Shift {
  @observable onDuty: User[];

  constructor(users: User[] = []) {
    this.onDuty = users;
  }

  @action
  addUser = (user: User) => {
    this.onDuty.push(user);
  };

  @action
  removeUser = (id: string) => {
    const index = this.onDuty.findIndex(user => user.id === id);
    this.onDuty.splice(index, 1);
  };

  toggleUser = (user: User) => {
    const hasUser = this.onDuty.some(({ id }) => id === user.id);

    if (hasUser) {
      this.removeUser(user.id);
    } else {
      this.addUser(user);
    }
  };

  fill = (users: User[]) => {
    const newUsersCount = USERS_PER_SHIFT - this.onDuty.length;
    const usersToAdd = users.slice(0, newUsersCount);

    for (let i = 0; i < usersToAdd.length; i += 1) {
      this.addUser(users[i]);
    }
  };

  @computed
  get isReady(): boolean {
    return this.onDuty.length === USERS_PER_SHIFT;
  }
}

export default Shift;
