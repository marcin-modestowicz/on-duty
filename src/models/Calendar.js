// @flow
import { observable, computed, action } from "mobx";
import type User from "./User";
import Availability, { AVAILABILITY_STATUSES } from "./Availability";
import Shift, { USERS_PER_SHIFT } from "./Shift";

export const MINIMUM_REST_DAYS = 1;

export function daysInCurrentMonth() {
  const currentDate = new Date();
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth() + 1;

  return new Date(year, month, 0).getDate();
}

export class ShiftsCalendar {
  @observable days: Shift[] = [];
  @observable summary: { [key: string]: string | number };

  constructor() {
    const maxDay = daysInCurrentMonth();

    for (let day = 1; day <= maxDay; day += 1) {
      this.days.push(new Shift());
    }
  }

  getUserShifts(userId: string): number[] {
    return this.days.reduce((sum, { isSealed, onDuty }, index) => {
      if (isSealed && onDuty.some(user => user.id === userId)) {
        sum.push(index);
      }
      return sum;
    }, []);
  }

  fillCalendar(users: User[]) {
    // Fill all slots in shift with users
    this.days.forEach((shift, index) => {
      users.forEach(user => {
        shift.addUser(user);
      });
    });

    // Create index array to iterate over, sort by preference
    const indexes = this.days
      .slice()
      .map((shift, index) => {
        return {
          index,
          preference: shift.onDuty.reduce(
            (sum, user) =>
              sum +
              (user.availabilityCalendar.days[index].status ===
              AVAILABILITY_STATUSES.KEEN
                ? user.power
                : 0),
            0
          )
        };
      })
      .sort((shiftA, shiftB) => {
        return shiftB.preference - shiftA.preference;
      })
      .map(shift => shift.index);

    // Iterate over shifts, sort users in shift
    indexes.forEach(index => {
      const currentShift = this.days[index];
      const notAllLimitsReached = users.some(
        user => this.getUserShifts(user.id).length < user.requiredShifts
      );

      // Sort users in shift
      currentShift._onDuty = currentShift._onDuty.sort((userA, userB) => {
        const userASortValue = this.getUserSortValue(
          userA,
          index,
          notAllLimitsReached
        );
        const userBSortValue = this.getUserSortValue(
          userB,
          index,
          notAllLimitsReached
        );

        return userASortValue - userBSortValue;
      });

      // Seal shift
      currentShift.seal();
    });

    // Create summary
    this.summary = this.createSummary(users);
  }

  getUserSortValue(
    user: User,
    shiftIndex: number,
    notAllLimitsReached: boolean
  ): number {
    const userShifts = this.getUserShifts(user.id);
    const userShiftCount = userShifts.length;
    const hasUserMinimumRestDays = !(
      userShifts.some(
        index => shiftIndex >= index - MINIMUM_REST_DAYS && shiftIndex < index
      ) ||
      userShifts.some(
        index => shiftIndex <= index + MINIMUM_REST_DAYS && shiftIndex > index
      )
    );
    const hasUserRequiredShiftsFilled = userShiftCount === user.requiredShifts;
    const userAvailabilityStatus =
      user.availabilityCalendar.days[shiftIndex].status;

    let userSortValue = -userAvailabilityStatus * user.power;

    if (!hasUserMinimumRestDays) {
      userSortValue += 20;
    }

    if (hasUserRequiredShiftsFilled && notAllLimitsReached) {
      userSortValue += 1;
    }

    return userSortValue;
  }

  createSummary(users: User[]) {
    const summary = users.reduce((allUsers, currentUser, index) => {
      const availabilityDays = currentUser.availabilityCalendar.days;
      const shiftDays = this.getUserShifts(currentUser.id);
      const highPreferenceDays = availabilityDays.reduce((sum, day, index) => {
        if (day.status === AVAILABILITY_STATUSES.KEEN) {
          sum.push(index);
        }

        return sum;
      }, []);
      const highPreference = highPreferenceDays.length;
      const highPreferenceFilled = highPreferenceDays.filter(index =>
        shiftDays.includes(index)
      ).length;
      const lowPreferenceDays = availabilityDays.reduce((sum, day, index) => {
        if (day.status === AVAILABILITY_STATUSES.BUSY) {
          sum.push(index);
        }

        return sum;
      }, []);
      const lowPreference = lowPreferenceDays.length;
      const lowPreferenceFilled = lowPreferenceDays.filter(index =>
        shiftDays.includes(index)
      ).length;
      const satisfaction =
        (highPreferenceFilled === 0
          ? 0
          : highPreferenceFilled / highPreference) -
        (lowPreferenceFilled === 0 ? 0 : lowPreferenceFilled / lowPreference);

      allUsers[currentUser.id] = {
        id: currentUser.id,
        name: currentUser.name,
        power: currentUser.power,
        highPreference,
        highPreferenceFilled,
        lowPreference,
        lowPreferenceFilled,
        shifts: shiftDays.length,
        shiftsShare: `${shiftDays.length /
          (this.days.length * USERS_PER_SHIFT / users.length) *
          100}%`,
        satisfaction
      };

      return allUsers;
    }, {});

    return summary;
  }
}

export class AvailabilityCalendar {
  days: Availability[] = [];

  constructor(availability?: number[] = []) {
    const maxDay = availability.length
      ? availability.length
      : daysInCurrentMonth();

    for (let day = 1; day <= maxDay; day += 1) {
      // $FlowFixMe
      this.days.push(new Availability(availability[day - 1]));
    }
  }
}
