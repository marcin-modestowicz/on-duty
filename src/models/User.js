// @flow
import { AvailabilityCalendar } from "./Calendar";
import { type AvailabilityStatus } from "./Availability";
import uuid from "uuid/v1";

export const POWER = {
  RESIDENT: 0,
  DOCTOR: 1,
  SPECIALIST: 2
};

const REQUIRED_SHIFTS = {
  RESIDENT: 2,
  DOCTOR: 1,
  SPECIALIST: 1
};

class User {
  id: string;
  name: string;
  isDoctor: boolean;
  isSpecialist: boolean;
  isAdmin: boolean;
  availabilityCalendar: AvailabilityCalendar;

  constructor(
    name: string,
    isDoctor?: boolean,
    isSpecialist?: boolean,
    availability?: AvailabilityStatus[]
  ) {
    this.id = uuid();
    this.name = name;
    this.isDoctor = !!isDoctor;
    this.isSpecialist = !!isSpecialist;
    this.isAdmin = false;
    this.availabilityCalendar = new AvailabilityCalendar(availability);
  }

  get power(): number {
    let power = POWER.RESIDENT;

    if (this.isDoctor) {
      power += POWER.DOCTOR;
    }

    if (this.isSpecialist) {
      power += POWER.SPECIALIST;
    }

    return power;
  }

  get requiredShifts(): number {
    let requiredShifts = REQUIRED_SHIFTS.RESIDENT;

    if (this.isDoctor) {
      requiredShifts = REQUIRED_SHIFTS.DOCTOR;
    }

    if (this.isSpecialist) {
      requiredShifts = REQUIRED_SHIFTS.SPECIALIST;
    }

    return requiredShifts;
  }
}

export default User;
