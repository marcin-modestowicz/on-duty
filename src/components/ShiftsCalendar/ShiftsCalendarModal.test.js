// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Shift from "../../models/Shift";
import User from "../../models/User";
import { AVAILABILITY_STATUSES } from "../../models/Availability";
import ShiftsCalendarModal from "./ShiftsCalendarModal";

describe("ShiftsCalendarModal component", () => {
  let RealDate;
  let shiftsCalendarModal;

  beforeAll(() => {
    RealDate = Date;
    const date = new Date(2017, 9, 1);
    global.Date = function DateMock(...args) {
      return args.length ? new RealDate(...args) : date;
    };

    const users = [
      new User("user1", false, false),
      new User("user2", false, false, [AVAILABILITY_STATUSES.BUSY]),
      new User("user3", true, false, [AVAILABILITY_STATUSES.KEEN])
    ];
    users.forEach((user, index) => {
      user.id = `a${index}`;
    });
    const day = {
      date: new Date(2017, 10, 1),
      shift: new Shift([users[0]])
    };

    shiftsCalendarModal = shallow(
      <ShiftsCalendarModal day={day} users={users} onClose={jest.fn()} />
    );
  });

  afterAll(() => {
    global.Date = RealDate;
  });

  test("should render", () => {
    expect(toJson(shiftsCalendarModal)).toMatchSnapshot();
  });
});
