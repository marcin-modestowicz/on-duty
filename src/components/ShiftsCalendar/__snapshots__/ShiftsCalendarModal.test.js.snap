// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftsCalendarModal component should render 1`] = `
<div
  className="root"
>
  <div
    className="modal"
  >
    <div
      className="header"
    >
      <Button
        marginless={true}
        onClick={[Function]}
      >
        Close
      </Button>
    </div>
    <div
      className="content"
    >
      <CheckboxButton
        checked={true}
        color="yellow"
        id="a0"
        key="a0"
        label="user1"
        onChange={[Function]}
        value="a0"
      />
      <CheckboxButton
        checked={false}
        color="red"
        id="a1"
        key="a1"
        label="user2"
        onChange={[Function]}
        value="a1"
      />
      <CheckboxButton
        checked={false}
        color="green"
        id="a2"
        key="a2"
        label="user3"
        onChange={[Function]}
        value="a2"
      />
    </div>
  </div>
</div>
`;
